plugins/outofbandmanagement-drivers/ipmitool/test/org/apache/cloudstack/outofbandmanagement/driver/ipmitool/IpmitoolWrapperTest.java

```diff
@@ -0,0 +1,107 @@
+        Assert.assertEquals(IpmitoolWrapper.parsePowerCommand(OutOfBandManagement.PowerOperation.OFF), "off");
+        Assert.assertEquals(IpmitoolWrapper.parsePowerCommand(OutOfBandManagement.PowerOperation.CYCLE), "cycle");
+        Assert.assertEquals(IpmitoolWrapper.parsePowerCommand(OutOfBandManagement.PowerOperation.RESET), "reset");
+        Assert.assertEquals(IpmitoolWrapper.parsePowerCommand(OutOfBandManagement.PowerOperation.SOFT), "soft");
+        Assert.assertEquals(IpmitoolWrapper.parsePowerCommand(OutOfBandManagement.PowerOperation.STATUS), "status");
+    }
+
+    @Test
+    public void testParsePowerCommandInvalid() {
+        try {
+            IpmitoolWrapper.parsePowerCommand(null);
+            Assert.fail("IpmitoolWrapper.parsePowerCommand failed to throw exception on invalid power state");
+        } catch (IllegalStateException e) {
+        }
+    }
+
+    @Test
+    public void testParsePowerState() {
+        Assert.assertEquals(IpmitoolWrapper.parsePowerState(null), OutOfBandManagement.PowerState.Unknown);
+        Assert.assertEquals(IpmitoolWrapper.parsePowerState(""), OutOfBandManagement.PowerState.Unknown);
+        Assert.assertEquals(IpmitoolWrapper.parsePowerState(" "), OutOfBandManagement.PowerState.Unknown);
+        Assert.assertEquals(IpmitoolWrapper.parsePowerState("invalid data"), OutOfBandManagement.PowerState.Unknown);
+        Assert.assertEquals(IpmitoolWrapper.parsePowerState("Chassis Power is on"), OutOfBandManagement.PowerState.On);
+        Assert.assertEquals(IpmitoolWrapper.parsePowerState("Chassis Power is off"), OutOfBandManagement.PowerState.Off);
+    }
+
+    @Test
+    public void testGetIpmiToolCommandArgs() {
+        List<String> args = IpmitoolWrapper.getIpmiToolCommandArgs("binpath", "intf", "1", null);
+        assert args != null;
+        Assert.assertEquals(args.size(), 6);
+        Assert.assertArrayEquals(args.toArray(), new String[]{"binpath", "-I", "intf", "-R", "1", "-v"});
+
+        ImmutableMap.Builder<OutOfBandManagement.Option, String> argMap = new ImmutableMap.Builder<>();
+        argMap.put(OutOfBandManagement.Option.DRIVER, "ipmitool");
+        argMap.put(OutOfBandManagement.Option.ADDRESS, "127.0.0.1");
+        List<String> argsWithOpts = IpmitoolWrapper.getIpmiToolCommandArgs("binpath", "intf", "1", argMap.build(), "user", "list");
+        assert argsWithOpts != null;
+        Assert.assertEquals(argsWithOpts.size(), 10);
+        Assert.assertArrayEquals(argsWithOpts.toArray(), new String[]{"binpath", "-I", "intf", "-R", "1", "-v", "-H", "127.0.0.1", "user", "list"});
+    }
+
+    @Test
+    public void testFindIpmiUser() {
+        // Invalid data
+        try {
+            IpmitoolWrapper.findIpmiUser("some invalid string\n", "admin");
+            Assert.fail("IpmitoolWrapper.findIpmiUser failed to throw exception on invalid data");
+        } catch (CloudRuntimeException ignore) {
+        }
```

Comment:
Consider splitting the success and failure test cases into separate methods with the failure method using the ``@Test(expected=CloudRuntimeException.class)`` annotation.